lib/knapsack_pro/extensions/rspec_extension.rb

```diff
@@ -0,0 +1,164 @@
+
+        def knapsack__seed_used?
+          configuration.seed_used?
+        end
+
+        # @param test_file_paths Array[String]
+        #   Example: ['a_spec.rb', 'b_spec.rb[1:1]']
+        def knapsack__load_spec_files_batch(test_file_paths)
+          world.reset
+
+          # Reset filters, but do not reset `configuration.static_config_filter_manager` to preserve the --tag option
+          filter_manager = RSpec::Core::FilterManager.new
+          options.configure_filter_manager(filter_manager)
+          configuration.filter_manager = filter_manager
+
+          configuration.knapsack__load_spec_files(test_file_paths)
+        end
+
+        # Based on:
+        # https://github.com/rspec/rspec-core/blob/f8c8880dabd8f0544a6f91d8d4c857c1bd8df903/lib/rspec/core/runner.rb#L113
+        #
+        # Option: `configuration.fail_if_no_examples`
+        #   Ignore the configuration.fail_if_no_examples option because it should be off in Queue Mode.
+        #   * Ignore the fail_if_no_examples option because in Queue Mode a late CI node can start after other CI nodes already executed tests. It is expected not to run examples in such a scenario.
+        #   * RSpec should not fail when examples are not executed for a batch of tests fetched from the Queue API. The batch could have tests that have no examples (for example, someone commented out the content of the spec file). We should fetch another batch of tests from the Queue API and keep running tests.
+        #
+        # @return [Fixnum] exit status code.
+        def knapsack__run_specs(queue_runner)
+          # Based on:
+          # https://github.com/rspec/rspec-core/blob/f8c8880dabd8f0544a6f91d8d4c857c1bd8df903/lib/rspec/core/world.rb#L53
+          ordering_strategy = configuration.ordering_registry.fetch(:global)
+          node_examples_passed = true
+
+          configuration.with_suite_hooks do
+            exit_status = configuration.reporter.report(_expected_example_count = 0) do |reporter|
+              queue_runner.with_batched_tests_from_queue do |test_file_paths|
+                knapsack__load_spec_files_batch(test_file_paths)
+
+                examples_passed = ordering_strategy.order(world.example_groups).map do |example_group|
+                  queue_runner.handle_signal!
+                  example_group.run(reporter)
+                end.all?
+
+                node_examples_passed = false unless examples_passed
+
+                knapsack__persist_example_statuses
+
+                if reporter.fail_fast_limit_met?
+                  queue_runner.log_fail_fast_limit_met
+                  break
+                end
+              end
+            end
+
+            exit_code(node_examples_passed)
+          end
+        end
+
+        # Based on:
+        # https://github.com/rspec/rspec-core/blob/f8c8880dabd8f0544a6f91d8d4c857c1bd8df903/lib/rspec/core/runner.rb#L90
+        def knapsack__persist_example_statuses
+          persist_example_statuses
```

Comment:
> I think you could inline this.

I decided not to do it because this method belongs to a different place in the original RSpec code, but we need to call it here after each batch of tests is executed. To make it more isolated I extracted the method to a separate `knapsack__` method with a comment that it is based on https://github.com/rspec/rspec-core/blob/f8c8880dabd8f0544a6f91d8d4c857c1bd8df903/lib/rspec/core/runner.rb#L90

Reply:
What's the value of this "isolation"?

If it's to point people at the source code, it could be inlined with a comment right above it.

But if you are strong about it, I'm ok with this.

Reply:
> it's to point people at the source code

Yes. In the original RSpec code the`persist_example_statuses` method is not called inside of `run_specs` but outside of it and in a different method `run`:
https://github.com/rspec/rspec-core/blob/f8c8880dabd8f0544a6f91d8d4c857c1bd8df903/lib/rspec/core/runner.rb#L89

I don't want to couple a few methods into one. I want isolation to make it easier to follow changes in original RSpec code if they happen.

> But if you are strong about it, I'm ok with this.

OK.